File: hbp100/schemas/entity.py
from dataclasses import dataclass, field
from typing import Optional, Dict, Any
from enum import Enum


class EntityType(str, Enum):
    """Types of entities that can be extracted."""

    # Personal Information
    NAME = "NAME"
    PHYSICIAN = "PHYSICIAN"
    PATIENT_ID = "PATIENT_ID"

    # Contact
    EMAIL = "EMAIL"
    PHONE = "PHONE"
    ADDRESS = "ADDRESS"

    # Medical Identifiers
    MRN = "MRN"
    CASE_ID = "CASE_ID"
    POLICY_NUMBER = "POLICY_NUMBER"

    # Dates
    DATE = "DATE"
    DOB = "DOB"

    # Other
    SSN = "SSN"
    HOSPITAL = "HOSPITAL"


@dataclass
class Entity:
    """
    Represents an extracted entity with all metadata.

    Attributes:
        type: Entity type (NAME, PHONE, EMAIL, etc.)
        value: The actual text value
        start: Character start position (for precise replacement)
        end: Character end position (for precise replacement)
        confidence: Extraction confidence (0.0-1.0)
        placeholder: Generated placeholder (set later)
        metadata: Additional entity-specific metadata
    """

    type: EntityType
    value: str
    start: int
    end: int
    confidence: float = 1.0
    placeholder: Optional[str] = None
    metadata: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        """Ensure end > start and valid confidence."""
        if self.end <= self.start:
            raise ValueError("End position must be greater than start")
        if not 0.0 <= self.confidence <= 1.0:
            raise ValueError("Confidence must be between 0.0 and 1.0")

    @property
    def text(self) -> str:
        """Alias for value."""
        return self.value

    @property
    def length(self) -> int:
        """Length of the entity value."""
        return len(self.value)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for serialization."""
        return {
            "type": self.type.value,
            "value": self.value,
            "start": self.start,
            "end": self.end,
            "confidence": self.confidence,
            "placeholder": self.placeholder,
            "metadata": self.metadata,
        }

    def __hash__(self) -> int:
        """Hash based on value and type."""
        return hash((self.type, self.value))

    def __eq__(self, other: object) -> bool:
        """Equality based on value and type."""
        if not isinstance(other, Entity):
            return False
        return self.type == other.type and self.value == other.value

File: src/schemas/entity.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Entity {
    pub entity_type: EntityType,
    pub value: String,
    pub start: usize,
    pub end: usize,
    pub confidence: f32,
    pub placeholder: Option<String>,
    pub metadata: HashMap<String, String>,
}

impl Entity {
    pub fn new(
        entity_type: EntityType,
        value: String,
        start: usize,
        end: usize,
        confidence: f32,
    ) -> Self {
        Self {
            entity_type,
            value,
            start,
            end,
            confidence,
            placeholder: None,
            metadata: HashMap::new(),
        }
    }

    pub fn to_dict(&self) -> HashMap<String, String> {
        let mut map = HashMap::new();
        map.insert("type".to_string(), self.entity_type.as_str().to_string());
        map.insert("value".to_string(), self.value.clone());
        map.insert("start".to_string(), self.start.to_string());
        map.insert("end".to_string(), self.end.to_string());
        map.insert("confidence".to_string(), self.confidence.to_string());
        map
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash, Copy)]
pub enum EntityType {
    Name,
    Email,
    Phone,
    Date,
    Address,
    Id,
    Medical,
}

impl From<String> for EntityType {
    fn from(s: String) -> Self {
        match s.to_uppercase().as_str() {
            "NAME" => EntityType::Name,
            "EMAIL" => EntityType::Email,
            "PHONE" => EntityType::Phone,
            "DATE" => EntityType::Date,
            "ADDRESS" => EntityType::Address,
            "ID" => EntityType::Id,
            "MEDICAL" => EntityType::Medical,
            _ => EntityType::Id,
        }
    }
}

impl EntityType {
    pub fn as_str(&self) -> &str {
        match self {
            EntityType::Name => "NAME",
            EntityType::Email => "EMAIL",
            EntityType::Phone => "PHONE",
            EntityType::Date => "DATE",
            EntityType::Address => "ADDRESS",
            EntityType::Id => "ID",
            EntityType::Medical => "MEDICAL",
        }
    }
}